File: linkedlist/142.c
/**
 * 仍旧利用快慢指针, 从第一次相遇点开始慢指针与头指针同步前进直到再次相遇 就是入口点
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *detectCycle(struct ListNode *head) {
    if(head ==NULL) return NULL;
    struct ListNode *fast, *slow;
    fast = head;
    slow = head;
    while(fast && fast->next)
    {
        fast = fast->next->next;
        slow = slow->next;
        
        if(fast == slow) break;
    }
    if(fast == NULL || fast->next ==NULL) return NULL;
    fast = head;
    while(fast)
    {
        if(fast == slow) break;
        fast = fast->next;
        slow = slow->next;
    }
    return fast;
    
}

File: linkedlist/142.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
    // fast slow node will meet before cycle node
    if head == nil {
        return nil
    }
    fast := head
    slow := head
    for fast != nil && fast.Next != nil {
        fast = fast.Next.Next
        slow = slow.Next
        if fast == slow {
            break
        }
    }
    if fast == nil || fast.Next == nil {
        return nil
    }
    fast = head
    for fast != nil {
        if fast == slow {
            break
        }
        fast = fast.Next
        slow = slow.Next
    }
    return fast
}
